horovod/spark/common/store.py

```diff
@@ -0,0 +1,334 @@
+        raise NotImplementedError()
+
+    def saving_runs(self):
+        raise NotImplementedError()
+
+    def get_runs_path(self):
+        raise NotImplementedError()
+
+    def get_run_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_logs_path(self, run_id):
+        raise NotImplementedError()
+
+    def get_checkpoint_filename(self):
+        raise NotImplementedError()
+
+    def get_logs_subdir(self):
+        raise NotImplementedError()
+
+    def exists(self, path):
+        raise NotImplementedError()
+
+    def read(self, path):
+        raise NotImplementedError()
+
+    def get_full_path_fn(self):
+        raise NotImplementedError()
+
+    def get_local_output_dir_fn(self, run_id):
+        raise NotImplementedError()
+
+    def sync_fn(self, run_id):
+        raise NotImplementedError()
+
+    def to_remote(self, run_id, dataset_idx):
+        attrs = self._remote_attrs(run_id, dataset_idx)
+
+        class RemoteStore(object):
+            def __init__(self):
+                for name, attr in attrs.items():
+                    setattr(self, name, attr)
+
+        return RemoteStore()
+
+    def _remote_attrs(self, run_id, dataset_idx):
+        return {
+            'train_data_path': self.get_train_data_path(dataset_idx),
+            'val_data_path': self.get_val_data_path(dataset_idx),
+            'test_data_path': self.get_test_data_path(dataset_idx),
+            'saving_runs': self.saving_runs(),
+            'runs_path': self.get_runs_path(),
+            'run_path': self.get_run_path(run_id),
+            'checkpoint_path': self.get_checkpoint_path(run_id),
+            'logs_path': self.get_logs_path(run_id),
+            'checkpoint_filename': self.get_checkpoint_filename(),
+            'logs_subdir': self.get_logs_subdir(),
+            'get_full_path': self.get_full_path_fn(),
+            'get_local_output_dir': self.get_local_output_dir_fn(run_id),
+            'sync': self.sync_fn(run_id)
+        }
+
+    @staticmethod
+    def create(prefix_path):
+        if HDFSStore.matches(prefix_path):
+            return HDFSStore(prefix_path)
+        else:
+            return LocalStore(prefix_path)
+
+
+class PrefixStore(Store):
+    def __init__(self, prefix_path, train_path=None, val_path=None, test_path=None, runs_path=None, save_runs=True):
+        self.prefix_path = self.get_normalized_path(prefix_path)
+        self._train_path = train_path or self._get_path('train_data')
+        self._val_path = val_path or self._get_path('val_data')
+        self._test_path = test_path or self._get_path('test_path')
+        self._runs_path = runs_path or self._get_path('runs')
+        self._save_runs = save_runs
+        super(PrefixStore, self).__init__()
+
+    def get_train_data_path(self, idx=None):
+        return '{}.{}'.format(self._train_path, idx) if idx is not None else self._train_path
```

Comment:
what about a more readable name? such as `{}_horovod_spark_estimator_intermediate_file_{}`

Reply:
Renamed to `intermediate_train_data` (etc.).  In most cases, the user should specify a path that is unique to their Horovod training (for output data like checkpoints), so I don't think we need to be too verbose.

Reply:
reasonable